vpr/src/analytical_place/flat_placement_density_manager.cpp

```diff
@@ -0,0 +1,302 @@
+                                  bin_region.top_right().x() - epsilon);
+    double y = std::clamp<double>(p_placement.block_y_locs[blk_id],
+                                  bin_region.bottom_left().y() + epsilon,
+                                  bin_region.top_right().y() - epsilon);
+    return vtr::Point<double>(x, y);
+}
+
+void FlatPlacementDensityManager::export_placement_from_bins(PartialPlacement& p_placement) const {
+    // Updates the partial placement with the location of the blocks in the bin
+    // by moiving the blocks to the point with the bin closest to where they
+    // were originally.
+    for (APBlockId blk_id : ap_netlist_.blocks()) {
+        // Only the moveable block locations should be exported.
+        if (ap_netlist_.block_mobility(blk_id) == APBlockMobility::FIXED)
+            continue;
+        // Project the coordinate of the block in the partial placement to the
+        // closest point in the bin.
+        FlatPlacementBinId blk_bin_id = bins_.block_bin(blk_id);
+        VTR_ASSERT_DEBUG(blk_bin_id.is_valid());
+        vtr::Point<double> new_blk_pos = get_block_location_in_bin(blk_id,
+                                                                   bins_.bin_region(blk_bin_id),
+                                                                   p_placement);
+        p_placement.block_x_locs[blk_id] = new_blk_pos.x();
+        p_placement.block_y_locs[blk_id] = new_blk_pos.y();
+        // NOTE: This code currently does not support 3D FPGAs.
+        VTR_ASSERT(std::floor(p_placement.block_layer_nums[blk_id]) == 0.0);
+    }
+}
+
+void FlatPlacementDensityManager::empty_bins() {
+    // Reset all of the bins and their utilizations.
+    for (FlatPlacementBinId bin_id : bins_.bins()) {
+        bins_.remove_all_blocks_from_bin(bin_id);
+        bin_utilization_[bin_id] = PrimitiveVector();
+        bin_overfill_[bin_id] = calc_bin_overfill(bin_utilization_[bin_id], bin_capacity_[bin_id]);
+        bin_underfill_[bin_id] = calc_bin_underfill(bin_utilization_[bin_id], bin_capacity_[bin_id]);
+    }
+    // Once all the bins are reset, all bins should be empty; therefore no bins
+    // are overfilled.
+    overfilled_bins_.clear();
+}
+
+bool FlatPlacementDensityManager::verify() const {
+    // Make sure that every block has a bin.
+    for (APBlockId blk_id : ap_netlist_.blocks()) {
+        if (!bins_.block_bin(blk_id).is_valid()) {
+            VTR_LOG("Bin Verify: Found a block that is not in a bin.\n");
+            return false;
+        }
+    }
+    // Make sure that every bin has the correct utilization, supply, and demand.
+    for (FlatPlacementBinId bin_id : bins_.bins()) {
+        PrimitiveVector calc_utilization;
+        for (APBlockId blk_id : bins_.bin_contained_blocks(bin_id)) {
+            calc_utilization += mass_calculator_.get_block_mass(blk_id);
+        }
+        if (bin_utilization_[bin_id] != calc_utilization) {
```

Comment:
Could add a verification routine to check that what blocks a bin thinks are placed in it matches what the blocks think in terms of what bins they are in.  That should go in a verify routine in the bin class itself.